src/wasm/wasm-type.cpp

```diff
@@ -45,300 +48,284 @@ struct TypeInfo {
-    {TypeInfo({Type::eqref}), Type::eqref},
-    {TypeInfo(HeapType(HeapType::EqKind), true), Type::eqref},
-    {TypeInfo({Type::i31ref}), Type::i31ref},
-    {TypeInfo(HeapType(HeapType::I31Kind), false), Type::i31ref},
-  };
+  // Track unique_ptrs for constructed types to avoid leaks.
+  std::vector<std::unique_ptr<Info>> constructedTypes;
 
-  uintptr_t canonicalize(const TypeInfo& info) {
+  // Maps from constructed types to their canonical Type IDs.
+  std::unordered_map<Info, uintptr_t> typeIDs;
+
+  typename Info::type_t canonicalize(Info info) {
     std::lock_guard<std::mutex> lock(mutex);
-    auto indexIt = indices.find(info);
-    if (indexIt != indices.end()) {
-      return indexIt->second;
+    auto indexIt = typeIDs.find(info);
+    if (indexIt != typeIDs.end()) {
+      return typename Info::type_t(indexIt->second);
     }
-    auto ptr = std::make_unique<TypeInfo>(info);
+    auto ptr = std::make_unique<Info>(info);
     auto id = uintptr_t(ptr.get());
     constructedTypes.push_back(std::move(ptr));
-    assert(id > Type::_last_basic_id);
-    indices[info] = id;
-    return id;
+    assert(id > Info::type_t::_last_basic_type);
+    typeIDs[info] = id;
+    return typename Info::type_t(id);
   }
+};
 
-  Type makeType(const Tuple& tuple) {
-    auto& types = tuple.types;
-#ifndef NDEBUG
-    for (Type t : types) {
-      assert(t.isSingle());
-    }
-#endif
-    if (types.size() == 0) {
-      return Type::none;
-    }
-    if (types.size() == 1) {
-      return types[0];
+struct TypeStore : Store<TypeInfo> {
+  Type canonicalize(TypeInfo info) {
+    if (info.isTuple()) {
+      if (info.tuple.types.size() == 0) {
+        return Type::none;
+      }
+      if (info.tuple.types.size() == 1) {
+        return info.tuple.types[0];
+      }
     }
-    return Type(canonicalize(TypeInfo(tuple)));
-  }
-
-  Type makeType(const HeapType& heapType, bool nullable) {
-#ifndef NDEBUG
-    switch (heapType.kind) {
-      case HeapType::FuncKind:
-      case HeapType::ExternKind:
-      case HeapType::AnyKind:
-      case HeapType::EqKind:
-      case HeapType::I31Kind:
-      case HeapType::ExnKind:
-      case HeapType::SignatureKind:
-        break;
-      case HeapType::StructKind:
-        for (Field f : heapType.struct_.fields) {
-          assert(f.type.isSingle());
+    if (info.isRef() && info.ref.heapType.isBasic()) {
+      if (info.ref.nullable) {
+        switch (info.ref.heapType.getBasic()) {
+          case HeapType::func:
+            return Type::funcref;
+          case HeapType::ext:
+            return Type::externref;
+          case HeapType::exn:
+            return Type::exnref;
+          case HeapType::any:
+            return Type::anyref;
+          case HeapType::eq:
+            return Type::eqref;
+          case HeapType::i31:
+            break;
         }
-        break;
-      case HeapType::ArrayKind:
-        assert(heapType.array.element.type.isSingle());
-        break;
+      } else {
```

Comment:
How about adding an assertion failure in case `heapType` is not nullable and is not `i31`, because we don't support it yet?

Reply:
I agree that erroring out when we use non-nullable types would make sense (including  for `i31ref`, which we don't support any better than other non-nullable types). But I think the validator would be a better place for that error since we support creating and querying non-nullable types, just not using them in the IR.

Reply:
What I was wondering is, even if the type is basic, if it is not nullable, it will fall out of all these `if`s and call `Store<TypeInfo>::canonicalize(info)`, which will create `TypeInfo` for it, in which case it can cause an unintended consequence which is not easy to catch later. Maybe I'm missing something?

Reply:
I don't think the act of making a `TypeInfo` for a non-nullable type can make anything bad happen. We also depend on being able to do that in test/example/typinfo.cpp.

The bad things only start happening once you use a non-nullable type for locals in the IR.

Reply:
But didn't you say you don't create `TypeInfo`s anymore with (nullable) basic types?

Reply:
Oh, I see what you're saying. We only provide `BasicType`s for types explicitly given their own names in the spec proposals, so `(ref null extern)`, which is equivalent to `externref`, is canonicalized into `Type::externref`, but on the other hand `(ref extern)` would be backed by a `TypeInfo`. This is working as intended. `i31ref` is weird because it is non-nullable by definition.